azurelinuxagent/common/utils/cryptutil.py

```diff
@@ -127,4 +140,14 @@ def bits_to_bytes(self, bits):
                 curr = 0
                 index = 7
         return bytes(byte_array)
-
+    
+    def decrypt_secret(self, encrypted_password, private_key, cache_file, encoding='utf-8'):
+        try:
+            decoded = textutil.b64decode(encrypted_password, encoding)
+            with open(cache_file, "wb") as c:
+                c.write(decoded)
+            decrypted_secret = self.decrypt_encrypted_file(private_key, cache_file)
+            return decrypted_secret.replace("\0", "")
```

Comment:
why is this necessary? please add comments

Reply:
This almost certainly means you're seeing unicode (16-bit chars) instead of utf-8, and just eliminating the NUL octets by hand is the wrong thing to do.

I assume you're testing all of this under 2.7 and 3.x, since the two pythons handle this very differently. As Hans and Chris have said, "Here Be Dragons."